manager/probe.go

```diff
@@ -532,46 +533,68 @@ func (p *Probe) reset() {
 	p.checkPin = false
 	p.funcName = ""
 	p.attachPID = 0
+	p.attachedWithSysfs = false
 }
 
 // attachKprobe - Attaches the probe to its kprobe
 func (p *Probe) attachKprobe() error {
 	// Prepare kprobe_events line parameters
-	var probeType, maxactiveStr string
+	var probeType, maxactiveStr, sectionPrefix string
 	var err error
 	funcName := p.funcName
-	if strings.HasPrefix(p.Section, "kretprobe/") {
+
+	dividedSection := strings.Split(p.Section, "/")
+	if len(dividedSection) >= 1 {
+		sectionPrefix = dividedSection[0]
+	} else {
+		return errors.Wrapf(ErrSectionFormat, "program type unrecognized in section %v", p.Section)
+	}
+
+	switch sectionPrefix {
+	case "kretprobe":
 		if funcName == "" {
 			funcName = strings.TrimPrefix(p.Section, "kretprobe/")
 		}
 		if p.KProbeMaxActive > 0 {
 			maxactiveStr = fmt.Sprintf("%d", p.KProbeMaxActive)
 		}
 		probeType = "r"
-	} else if strings.HasPrefix(p.Section, "kprobe/") {
+	case "kprobe":
 		if funcName == "" {
 			funcName = strings.TrimPrefix(p.Section, "kprobe/")
 		}
 		probeType = "p"
-	} else {
+	default:
 		// this might actually be a Uprobe
 		return p.attachUprobe()
 	}
 	p.attachPID = os.Getpid()
 
-	// Write kprobe_events line to register kprobe
-	kprobeID, err := EnableKprobeEvent(probeType, funcName, p.UID, maxactiveStr, p.attachPID)
-	// fallback without KProbeMaxActive
-	if err == ErrKprobeIDNotExist {
-		kprobeID, err = EnableKprobeEvent(probeType, funcName, p.UID, "", p.attachPID)
-	}
+	// Try to use the perf_event_open API first (e12f03d "perf/core: Implement the 'perf_kprobe' PMU")
+	perfEventFD, err := perfEventOpenWithProbe(funcName, 0, -1, sectionPrefix, 0)
 	if err != nil {
-		return errors.Wrapf(err, "couldn't enable kprobe %s", p.Section)
+		// Try to create the event using debugfs
+		// Write kprobe_events line to register kprobe
+		kprobeID, err := EnableKprobeEvent(probeType, funcName, p.UID, maxactiveStr, p.attachPID)
```

Comment:
The `perf_event_open` only method doesn't generate a `kprobeID` like `debugfs` does. I'm also not sure why this function is exported.

Reply:
Yeah, `EnableKprobeEvent` is by design the "old" ABI, so its only job is to write a line in `kprobe_events`. The `perf_event_open` happens after this stage with either the `kprobeID` or the returned fd of the new ABI. It's role is to activate the kprobe so that it triggers the eBPF program (just like when you write `1` in the `enable` file when you're using perf). So if that's ok with you both I can:
- rename `EnableKprobeEvent` to `RegisterKprobeEvent` (since it doesn't enable it per say ...)
- unexport the function

(I initially exported it because we used to write tests manually to check that the right probes were activated)

Reply:
(same goes for the `Disable` function &`*Uprobe` ones)